synapse/federation/federation_server.py

```diff
@@ -135,22 +141,41 @@ async def on_incoming_transaction(
         request_time = self._clock.time_msec()
 
         transaction = Transaction(**transaction_data)
+        transaction_id = transaction.transaction_id  # type: ignore
 
-        if not transaction.transaction_id:  # type: ignore
+        if not transaction_id:
             raise Exception("Transaction missing transaction_id")
 
-        logger.debug("[%s] Got transaction", transaction.transaction_id)  # type: ignore
+        logger.debug("[%s] Got transaction", transaction_id)
 
-        # use a linearizer to ensure that we don't process the same transaction
-        # multiple times in parallel.
+        # We wrap in a ResponseCache so that we de-duplicate retried
+        # transactions.
+        return await self._transaciton_resp_cache.wrap(
+            (origin, transaction_id),
+            self._on_incoming_transaction_inner,
+            origin,
+            transaction,
+            request_time,
+        )
+
+    async def _on_incoming_transaction_inner(
+        self, origin: str, transaction: Transaction, request_time: int
+    ) -> Tuple[int, Dict[str, Any]]:
+        # Use a linearizer to ensure that process transactions from a remote in
+        # order.
         with (
             await self._transaction_linearizer.queue(
                 (origin, transaction.transaction_id)  # type: ignore
             )
         ):
-            result = await self._handle_incoming_transaction(
-                origin, transaction, request_time
-            )
+            # We rate limit here *after* we've queued up the incoming requests,
+            # so that we don't fill up the ratelimiter with blocked requests.
+            with self._federation_ratelimiter.ratelimit(origin) as d:
+                await d
```

Comment:
I don't think that works as `ratelimit` returns a contextmanager which returns an awaitable, rather than `ratelimit` returning an awaitable that resolves to a contextmanager.

I kinda want to just rewrite this to either do it your way or use `async with`

Reply:
Ah, yeah that code is confusing. Seems like we should fix this in the future.